### PR TITLE
Create dao

### PR DESCRIPTION
**Pull request by MalishkaMyu**
last version of dao and dao tests

### CODE REVIEW

File: src/main/java/com/samsolutions/employeesdep/model/dao/JpaRoleDao.java [hunk 1]
```diff
@@ -0,0 +1,80 @@
...
+        Query query = entityManager.createQuery("SELECT r FROM Role r", Role.class);
+        return castRolesList(Role.class, query.getResultList());
+    }
+
+    @Override
+    @Transactional
+    public Long save(Role role) {
+        entityManager.persist(role);
+        Role savedRole = entityManager.find(Role.class, role.getId());
+        return savedRole.getId();
```
Comment (AlecKotovichSAM) [inline src/main/java/com/samsolutions/employeesdep/model/dao/JpaRoleDao.java]: check whether savedRole is not null (return null if it is null)

File: src/main/java/com/samsolutions/employeesdep/model/dao/JpaRoleDao.java [hunk 2]
```diff
@@ -0,0 +1,80 @@
...
+    @Override
+    @Transactional
+    public int deleteAll() {
+        Query query = entityManager.createQuery("DELETE from Role r");
+        return query.executeUpdate();
+    }
+
+    public static <Role> List<Role> castRolesList(Class<? extends Role> clazz, Collection<?> rawCollection) {
+        List<Role> result = new ArrayList<>(rawCollection.size());
+        for (Object o : rawCollection) {
```
Comment (AlecKotovichSAM) [inline src/main/java/com/samsolutions/employeesdep/model/dao/JpaRoleDao.java]: 0 -> role

Comment (AlecKotovichSAM) [inline src/main/java/com/samsolutions/employeesdep/model/dao/JpaRoleDao.java]: rawCollection -> roles

File: src/main/java/com/samsolutions/employeesdep/model/dao/JpaRoleDao.java [hunk 3]
```diff
@@ -0,0 +1,80 @@
...
+import javax.persistence.EntityManager;
+import javax.persistence.PersistenceContext;
+import javax.persistence.Query;
+import javax.transaction.Transactional;
+
+import org.springframework.stereotype.Repository;
+
+import com.samsolutions.employeesdep.model.entities.Role;
+
+@Repository
```
Comment (AlecKotovichSAM) [inline src/main/java/com/samsolutions/employeesdep/model/dao/JpaRoleDao.java:18]: Add @Slf4J

File: src/main/java/com/samsolutions/employeesdep/model/dao/JpaRoleDao.java [hunk 4]
```diff
@@ -0,0 +1,80 @@
...
+        return query.executeUpdate();
+    }
+
+    public static <Role> List<Role> castRolesList(Class<? extends Role> clazz, Collection<?> rawCollection) {
+        List<Role> result = new ArrayList<>(rawCollection.size());
+        for (Object o : rawCollection) {
+            try {
+                result.add(clazz.cast(o));
+            } catch (ClassCastException e) {
+                System.err.println("ClassCastException: " + e.getMessage());
```
Comment (AlecKotovichSAM) [inline src/main/java/com/samsolutions/employeesdep/model/dao/JpaRoleDao.java]: System.err.println -> log.error

File: src/main/java/com/samsolutions/employeesdep/model/entities/AbstractDateTimeEntity.java [hunk 5]
```diff
@@ -0,0 +1,26 @@
...
+public abstract class AbstractDateTimeEntity {
+    @Column(name = "created_at", nullable = false, updatable = false)
+    private LocalDateTime createdAt;
+
+    @Column(name = "updated_at", nullable = false)
+    private LocalDateTime updatedAt;
+
+    @PrePersist
+    protected void onCreate() {
+        createdAt = updatedAt = LocalDateTime.now();
```
Comment (AlecKotovichSAM) [inline src/main/java/com/samsolutions/employeesdep/model/entities/AbstractDateTimeEntity.java]: createdAt = LocalDateTime.now();
updatedAt = createdAt;

File: src/main/java/com/samsolutions/employeesdep/model/entities/Department.java [hunk 6]
```diff
@@ -3,11 +3,13 @@
 import org.hibernate.tuple.GenerationTiming;
 
 import javax.persistence.*;
```
Comment (AlecKotovichSAM) [inline src/main/java/com/samsolutions/employeesdep/model/entities/Department.java]: No imports with *

File: src/main/java/com/samsolutions/employeesdep/model/entities/Department.java [hunk 7]
```diff
@@ -3,11 +3,13 @@
...
+import java.io.Serializable;
+import java.util.List;
 import java.util.Set;
 
 @Entity
 @Table(name = "departments")
-public class Department {
+public class Department implements Serializable {
```
Comment (AlecKotovichSAM) [inline src/main/java/com/samsolutions/employeesdep/model/entities/Department.java:24]: @Getter?

File: src/main/java/com/samsolutions/employeesdep/model/entities/Employee.java [hunk 8]
```diff
@@ -2,14 +2,20 @@
 
 import com.samsolutions.employeesdep.model.converters.GenderConverter;
 import com.samsolutions.employeesdep.model.enums.Gender;
+import lombok.*;
 
 import javax.persistence.*;
```
Comment (AlecKotovichSAM) [inline src/main/java/com/samsolutions/employeesdep/model/entities/Employee.java]: same

File: src/main/java/com/samsolutions/employeesdep/model/entities/Employee.java [hunk 9]
```diff
@@ -35,88 +41,24 @@ public class Employee {
     @Column(name = "pass_valid")
     private LocalDate passValid;
```
Comment (AlecKotovichSAM) [inline src/main/java/com/samsolutions/employeesdep/model/entities/Employee.java]: rename to passValidity

File: src/main/java/com/samsolutions/employeesdep/model/entities/Employee.java [hunk 10]
```diff
@@ -2,14 +2,20 @@
...
+import java.util.Objects;
 import java.util.Set;
 
 @Entity
 @Table(name = "employees")
-public class Employee {
+@Getter
+@Setter
+@NoArgsConstructor
+public class Employee extends AbstractDateTimeEntity implements Serializable {
```
Comment (AlecKotovichSAM) [inline src/main/java/com/samsolutions/employeesdep/model/entities/Employee.java:33]: Add RequiredArgsConstructor annotation

File: src/main/java/com/samsolutions/employeesdep/model/entities/Employee.java [hunk 11]
```diff
@@ -35,88 +41,24 @@ public class Employee {
...
-    public void setId(Long id) {
-        this.id = id;
-    }
-
-    public String getName() {
-        return name;
-    }
-
-    public void setName(String name) {
+    public Employee(String name, String surname, Gender sex, LocalDate birthDate, LocalDate employmentDate) {
```
Comment (AlecKotovichSAM) [inline src/main/java/com/samsolutions/employeesdep/model/entities/Employee.java:76]: Mark  all required field with NonNull annotation

Comment (AlecKotovichSAM) [reply src/main/java/com/samsolutions/employeesdep/model/entities/Employee.java:76]: And remove this constructor afterwards

File: src/main/java/com/samsolutions/employeesdep/model/entities/Employee.java [hunk 12]
```diff
@@ -35,88 +41,24 @@ public class Employee {
...
-        this.department = department;
-    }
-
-    public Set<Role> getEmployeeRoles() {
-        return employeeRoles;
-    }
-
-    public void setEmployeeRoles(Set<Role> employeeRoles) {
-        this.employeeRoles = employeeRoles;
+        this.employmentDate = Objects.requireNonNullElse(employmentDate, LocalDate.now());
```
Comment (AlecKotovichSAM) [inline src/main/java/com/samsolutions/employeesdep/model/entities/Employee.java:81]: like

File: src/main/java/com/samsolutions/employeesdep/model/entities/Role.java [hunk 13]
```diff
@@ -18,6 +18,9 @@ public class Role {
     public Role() {
     }
 
+    public Role(String role) {
```
Comment (AlecKotovichSAM) [inline src/main/java/com/samsolutions/employeesdep/model/entities/Role.java]: Add lombok

File: src/main/java/com/samsolutions/employeesdep/model/repository/DepartmentRepository.java [hunk 14]
```diff
@@ -0,0 +1,16 @@
...
+import org.springframework.data.jpa.repository.JpaRepository;
+import org.springframework.data.jpa.repository.Query;
+import org.springframework.data.repository.query.Param;
+import org.springframework.stereotype.Repository;
+import org.springframework.transaction.annotation.Transactional;
+
+import java.util.List;
+
+@Repository
+public interface DepartmentRepository extends JpaRepository<Department,Long> {
```
Comment (AlecKotovichSAM) [inline src/main/java/com/samsolutions/employeesdep/model/repository/DepartmentRepository.java]: Non-formatted code

File: src/main/java/com/samsolutions/employeesdep/model/repository/EmployeeRepository.java [hunk 15]
```diff
@@ -0,0 +1,24 @@
...
+@Repository
+public interface EmployeeRepository extends JpaRepository<Employee, Long> {
+    @Query("select e from Employee e where e.name = :name and e.surname = :surname")
+    Employee findByName(@Param("name") String name, @Param("surname") String surname);
+
+    @Query("select e from Employee e join e.department d where d.name = :department order by e.surname, e.name")
+    List<Employee> findInDepartment(@Param("department") String department);
+
+    @Query("select count(r) from Employee e left join e.employeeRoles r where e.id = :emp_id")
+    //@Query("select SIZE(e.employeeRoles) from Employee e where e.id = :emp_id")
```
Comment (AlecKotovichSAM) [inline src/main/java/com/samsolutions/employeesdep/model/repository/EmployeeRepository.java]: Remove commented line

File: src/main/java/com/samsolutions/employeesdep/model/services/EmployeeService.java [hunk 16]
```diff
@@ -0,0 +1,41 @@
...
+
+    @Autowired
+    private JpaRoleDao roleRepository;
+
+    public List<Employee> getEmployeesToDepartment(String departName) {
+        if (departName.equals(""))
+            return empRepository.findAll();
+        else
+            return empRepository.findInDepartment(departName);
+    }
```
Comment (AlecKotovichSAM) [inline src/main/java/com/samsolutions/employeesdep/model/services/EmployeeService.java:32]: if (departName.equals("")) {
          return empRepository.findAll();
} else {
          return empRepository.findInDepartment(departName);
}

File: src/main/java/com/samsolutions/employeesdep/model/services/EmployeeService.java [hunk 17]
```diff
@@ -0,0 +1,41 @@
...
+    private EmployeeRepository empRepository;
+
+    @Autowired
+    private DepartmentRepository departRepository;
+
+    @Autowired
+    private JpaRoleDao roleRepository;
+
+    public List<Employee> getEmployeesToDepartment(String departName) {
+        if (departName.equals(""))
```
Comment (AlecKotovichSAM) [inline src/main/java/com/samsolutions/employeesdep/model/services/EmployeeService.java]: departName.equals("") -> StringUtils.isBlank(departName)

null
""
"   "

File: src/main/java/com/samsolutions/employeesdep/model/services/EmployeeService.java (hunk 16 above)
Comment (AlecKotovichSAM) [reply src/main/java/com/samsolutions/employeesdep/model/services/EmployeeService.java:32]: StringUtils.isBlank(departName)

File: src/test/java/com/samsolutions/employeesdep/model/dao/JpaRoleDaoTest.java [hunk 18]
```diff
@@ -0,0 +1,63 @@
...
+
+import java.util.List;
+
+import static org.junit.jupiter.api.Assertions.assertEquals;
+
+@SpringBootTest
+public class JpaRoleDaoTest {
+    @Autowired
+    private JpaRoleDao jpaRoleDao;
+
```
Comment (AlecKotovichSAM) [inline src/test/java/com/samsolutions/employeesdep/model/dao/JpaRoleDaoTest.java]: remove empty line

File: src/test/java/com/samsolutions/employeesdep/model/dao/JpaRoleDaoTest.java [hunk 19]
```diff
@@ -0,0 +1,63 @@
...
+        readRoles = jpaRoleDao.findAll();
+        assertEquals(1, readRoles.size());
+        assertEquals("Programmer", readRoles.get(0).getRole());
+    }
+
+    @AfterEach
+    public void tearDown() {
+        jpaRoleDao.deleteById(this.id);
+        assertEquals(0, jpaRoleDao.findAll().size());
+    }
```
Comment (AlecKotovichSAM) [inline src/test/java/com/samsolutions/employeesdep/model/dao/JpaRoleDaoTest.java:73]: import static org.hamcrest.MatcherAssert.assertThat;
import static org.hamcrest.CoreMatchers.is;
import static org.hamcrest.Matchers.empty;

assertThat(jpaRoleDao.findAll(), is(empty()));

File: src/test/java/com/samsolutions/employeesdep/repository/DepartmentRepositoryTest.java [hunk 20]
```diff
@@ -0,0 +1,64 @@
...
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.boot.test.context.SpringBootTest;
+
+import org.junit.jupiter.api.Test;
+import org.springframework.test.annotation.Rollback;
+
+import javax.transaction.Transactional;
+import java.util.List;
+
+import static org.junit.jupiter.api.Assertions.*;
```
Comment (AlecKotovichSAM) [inline src/test/java/com/samsolutions/employeesdep/repository/DepartmentRepositoryTest.java]: import *

File: src/test/java/com/samsolutions/employeesdep/repository/DepartmentRepositoryTest.java [hunk 21]
```diff
@@ -0,0 +1,64 @@
...
+        repository.save(depart1);
+        assertEquals("PHP Department", depart1.getName());
+        assertNotEquals("Java Department", depart1.getName());
+    }
+
+    @AfterEach
+    public void tearDown() {
+        repository.deleteById(depart1.getId());
+        repository.delete(depart2);
+        assertEquals(0, repository.findAll().size());
```
Comment (AlecKotovichSAM) [inline src/test/java/com/samsolutions/employeesdep/repository/DepartmentRepositoryTest.java]: assertThat(repository.findAll(), is(empty()));

File: src/test/java/com/samsolutions/employeesdep/repository/EmployeeRepositoryTest.java [hunk 22]
```diff
@@ -0,0 +1,196 @@
...
+            assertEquals("Java Department", readEmp.getDepartment().getName());
+            assertEquals(3, repository.countRolesOfEmployee(readEmp.getId()));
+        }
+    }
+
+    @Test
+    public void testCreateEmpInDepartmentDelete() {
+        // creating the new employee with the existing department
+        Employee empNew = new Employee("Nyusha", "Smesharik", Gender.FEMALE,
+                LocalDate.of(2010, 1, 13), null);
```
Comment (AlecKotovichSAM) [inline src/test/java/com/samsolutions/employeesdep/repository/EmployeeRepositoryTest.java:80]: Introduce local var

File: src/test/java/com/samsolutions/employeesdep/repository/EmployeeRepositoryTest.java [hunk 23]
```diff
@@ -0,0 +1,196 @@
...
+    @Test
+    public void testReadOneUpdateNewDepartment() {
+        //reading of the existing employee "Krosh Smescharik"
+        Employee readUpdEmp = repository.findById(emp1.getId()).get();
+        assertEquals("Smesharik", readUpdEmp.getSurname());
+        // update name and pass number
+        readUpdEmp.setName("Losyash");
+        readUpdEmp.setPassNumber("SH123456");
+        // setting the new roles - 2 old
+        Set<Role> empRoles = new HashSet<>();
```
Comment (AlecKotovichSAM) [inline src/test/java/com/samsolutions/employeesdep/repository/EmployeeRepositoryTest.java:115]: HashSet<>(2)

File: src/test/java/com/samsolutions/employeesdep/repository/EmployeeRepositoryTest.java [hunk 24]
```diff
@@ -0,0 +1,196 @@
...
+
+    @AfterEach
+    public void tearDown() {
+        repository.delete(emp1);
+        assertEquals(0, repository.findAll().size());
+        assertEquals(1, departRepository.findAll().size());
+        assertEquals(3, roleRepository.findAll().size());
+        departRepository.delete(depart1);
+        assertEquals(0, departRepository.findAll().size());
+        /*for (Role role : roles) {
```
Comment (AlecKotovichSAM) [inline src/test/java/com/samsolutions/employeesdep/repository/EmployeeRepositoryTest.java]: remove commented code

**Review by AlecKotovichSAM — CHANGES_REQUESTED** [review]

File: src/main/java/com/samsolutions/employeesdep/model/dao/JpaRoleDao.java (hunk 1 above)
Comment (MalishkaMyu) [reply src/main/java/com/samsolutions/employeesdep/model/dao/JpaRoleDao.java]: done

File: src/main/java/com/samsolutions/employeesdep/model/dao/JpaRoleDao.java (hunk 2 above)
Comment (MalishkaMyu) [reply src/main/java/com/samsolutions/employeesdep/model/dao/JpaRoleDao.java]: done

Comment (MalishkaMyu) [reply src/main/java/com/samsolutions/employeesdep/model/dao/JpaRoleDao.java]: done

File: src/main/java/com/samsolutions/employeesdep/model/dao/JpaRoleDao.java (hunk 4 above)
Comment (MalishkaMyu) [reply src/main/java/com/samsolutions/employeesdep/model/dao/JpaRoleDao.java]: done

File: src/main/java/com/samsolutions/employeesdep/model/dao/JpaRoleDao.java (hunk 3 above)
Comment (MalishkaMyu) [reply src/main/java/com/samsolutions/employeesdep/model/dao/JpaRoleDao.java:18]: done

File: src/test/java/com/samsolutions/employeesdep/repository/DepartmentRepositoryTest.java (hunk 20 above)
Comment (MalishkaMyu) [reply src/test/java/com/samsolutions/employeesdep/repository/DepartmentRepositoryTest.java]: done

File: src/test/java/com/samsolutions/employeesdep/repository/DepartmentRepositoryTest.java (hunk 21 above)
Comment (MalishkaMyu) [reply src/test/java/com/samsolutions/employeesdep/repository/DepartmentRepositoryTest.java]: done

File: src/test/java/com/samsolutions/employeesdep/repository/EmployeeRepositoryTest.java (hunk 22 above)
Comment (MalishkaMyu) [reply src/test/java/com/samsolutions/employeesdep/repository/EmployeeRepositoryTest.java:80]: done

File: src/test/java/com/samsolutions/employeesdep/repository/EmployeeRepositoryTest.java (hunk 23 above)
Comment (MalishkaMyu) [reply src/test/java/com/samsolutions/employeesdep/repository/EmployeeRepositoryTest.java:115]: done (1)

File: src/test/java/com/samsolutions/employeesdep/repository/EmployeeRepositoryTest.java (hunk 24 above)
Comment (MalishkaMyu) [reply src/test/java/com/samsolutions/employeesdep/repository/EmployeeRepositoryTest.java]: done

File: src/main/java/com/samsolutions/employeesdep/model/entities/AbstractDateTimeEntity.java (hunk 5 above)
Comment (MalishkaMyu) [reply src/main/java/com/samsolutions/employeesdep/model/entities/AbstractDateTimeEntity.java]: done

File: src/main/java/com/samsolutions/employeesdep/model/entities/Department.java (hunk 6 above)
Comment (MalishkaMyu) [reply src/main/java/com/samsolutions/employeesdep/model/entities/Department.java]: done

File: src/main/java/com/samsolutions/employeesdep/model/entities/Department.java (hunk 7 above)
Comment (MalishkaMyu) [reply src/main/java/com/samsolutions/employeesdep/model/entities/Department.java:24]: done

File: src/main/java/com/samsolutions/employeesdep/model/entities/Employee.java (hunk 8 above)
Comment (MalishkaMyu) [reply src/main/java/com/samsolutions/employeesdep/model/entities/Employee.java]: done

File: src/main/java/com/samsolutions/employeesdep/model/entities/Employee.java (hunk 9 above)
Comment (MalishkaMyu) [reply src/main/java/com/samsolutions/employeesdep/model/entities/Employee.java]: done

File: src/main/java/com/samsolutions/employeesdep/model/entities/Employee.java (hunk 10 above)
Comment (MalishkaMyu) [reply src/main/java/com/samsolutions/employeesdep/model/entities/Employee.java:33]: 1 field in my constructor is optional and comouted if null - not done

File: src/main/java/com/samsolutions/employeesdep/model/entities/Employee.java (hunk 11 above)
Comment (MalishkaMyu) [reply src/main/java/com/samsolutions/employeesdep/model/entities/Employee.java:76]: 1 field in my constructor is optional and comouted if null - not done

File: src/main/java/com/samsolutions/employeesdep/model/entities/Role.java (hunk 13 above)
Comment (MalishkaMyu) [reply src/main/java/com/samsolutions/employeesdep/model/entities/Role.java]: done

File: src/main/java/com/samsolutions/employeesdep/model/repository/DepartmentRepository.java (hunk 14 above)
Comment (MalishkaMyu) [reply src/main/java/com/samsolutions/employeesdep/model/repository/DepartmentRepository.java]: done

File: src/main/java/com/samsolutions/employeesdep/model/services/EmployeeService.java (hunk 16 above)
Comment (MalishkaMyu) [reply src/main/java/com/samsolutions/employeesdep/model/services/EmployeeService.java:32]: done

File: src/main/java/com/samsolutions/employeesdep/model/repository/EmployeeRepository.java (hunk 15 above)
Comment (MalishkaMyu) [reply src/main/java/com/samsolutions/employeesdep/model/repository/EmployeeRepository.java]: done

File: src/main/java/com/samsolutions/employeesdep/model/services/EmployeeService.java (hunk 17 above)
Comment (MalishkaMyu) [reply src/main/java/com/samsolutions/employeesdep/model/services/EmployeeService.java]: done

File: src/test/java/com/samsolutions/employeesdep/model/dao/JpaRoleDaoTest.java (hunk 18 above)
Comment (MalishkaMyu) [reply src/test/java/com/samsolutions/employeesdep/model/dao/JpaRoleDaoTest.java]: done

File: src/test/java/com/samsolutions/employeesdep/model/dao/JpaRoleDaoTest.java (hunk 19 above)
Comment (MalishkaMyu) [reply src/test/java/com/samsolutions/employeesdep/model/dao/JpaRoleDaoTest.java:73]: done

**Comment by MalishkaMyu** [comment]
Hi Oleg. 
Ich fixed everything.
Could you check and merge please?

**Review by AlecKotovichSAM — APPROVED** [review]

**Pull request merged**